components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,240 @@ package xid
+					continue
+				}
+				c.Lock()
+				c.currState = EvolveHealthyState(events)
+				c.Unlock()
+			case dmesgLine := <-watcher.Watch():
+				log.Logger.Debugw("dmesg line", "line", dmesgLine)
+				ev, err := nvidia_query_xid.ParseDmesgLogLine(metav1.Time{Time: dmesgLine.Timestamp}, dmesgLine.Content)
+				if err != nil {
+					log.Logger.Errorw("failed to parse dmesg line", "error", err)
+					continue
+				}
+				if ev.Detail == nil {
+					log.Logger.Debugw("not xid event, skip")
+					continue
+				}
+				event := components.Event{
+					Time: metav1.Time{Time: dmesgLine.Timestamp},
+					Name: EventNameErroXid,
+					ExtraInfo: map[string]string{
+						EventKeyErroXidData: strconv.FormatInt(int64(ev.Detail.Xid), 10),
+						EventKeyDeviceUUID:  ev.DeviceUUID,
+					},
+				}
+				currEvent, err := c.store.Find(c.rootCtx, event)
+				if err != nil {
+					log.Logger.Errorw("failed to create event", "error", err)
+					continue
+				}
+				if currEvent != nil {
+					log.Logger.Debugw("no new events created")
+					continue
+				}
+				if err = c.store.Insert(c.rootCtx, event); err != nil {
+					log.Logger.Errorw("failed to create event", "error", err)
+					continue
+				}
+				events, err := c.store.Get(c.rootCtx, time.Time{})
+				if err != nil {
+					log.Logger.Errorw("failed to get all events", "error", err)
+					continue
+				}
+				c.Lock()
+				c.currState = EvolveHealthyState(events)
+				c.Unlock()
+			}
 		}
+	}()
+	return nil
+}
 
-		lastSuccessPollElapsed := time.Now().UTC().Sub(ev.Time.Time)
-		if lastSuccessPollElapsed > 2*c.poller.Config().Interval.Duration {
-			log.Logger.Warnw("last poll is too old", "elapsed", lastSuccessPollElapsed, "interval", c.poller.Config().Interval.Duration)
-		}
-	}
-
-	return o, nil
+func (c *XIDComponent) States(_ context.Context) ([]components.State, error) {
+	c.RLock()
+	defer c.RUnlock()
+	return []components.State{c.currState}, nil
 }
 
-func (c *component) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
-	o, err := c.tailScan()
+func (c *XIDComponent) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
+	var ret []components.Event
+	events, err := c.store.Get(ctx, since)
 	if err != nil {
 		return nil, err
 	}
-	return o.getEvents(since), nil
+	for _, event := range events {
+		ret = append(ret, resolveXIDEvent(event))
+	}
+	return ret, nil
 }
 
-func (c *component) Metrics(ctx context.Context, since time.Time) ([]components.Metric, error) {
+func (c *XIDComponent) Metrics(ctx context.Context, since time.Time) ([]components.Metric, error) {
 	log.Logger.Debugw("querying metrics", "since", since)
 
 	return nil, nil
 }
 
-func (c *component) Close() error {
-	log.Logger.Debugw("closing component")
-
+func (c *XIDComponent) Close() error {
+	log.Logger.Debugw("closing XIDComponent")
 	// safe to call stop multiple times
 	c.poller.Stop(nvidia_component_error_xid_id.Name)
-
+	c.cancel()
```

Comment:
I'm not sure if cancel is only called once. If not, we should use sync.Once to ensure it's only called once.

Reply:
yea its only called once (since component.Close will only be called once)